Added processing directory usage page. refs #7921

mcantelon opened the pull request:
Adds an admin page where processing space usage is shown and these processing areas can be purged.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+    shared_dir = config.get('MCPClient', 'sharedDirectoryMounted')
+
+    # Get DIP uploads directory
+    dip_upload_dir = os.path.join(shared_dir, 'watchedDirectories/uploadDIP')
```
This should be uploadedDIPs.  uploadDIP is the watched directory where we ask where the DIP should be uploaded to.  uploadedDIPs is where DIPs that have already been uploaded go to.  Also, splitting the path components given to os.path.join is preferred.

mcantelon replied:
So, for the splitting, something like this?

```
path_parts = [os.path.sep] + shared_dir.split('/') + 'watchedDirectories/uploadDIP'.split('/')
dip_upload_dir = os.path.join(*path_parts)
```

Hwesta replied:
No, I meant that having a literal `/` in a parameter to os.path.join kinda defeats the purpose, since os.path.join behaves differently depending on the OS and `/` pins it to a particular OS.  Should be `os.path.join(shared_dir, 'watchedDirectories', 'uploadedDIPs')`.  Here I'm also assuming shared_dir is constructed with os.path.join (or read from a config file with an OS-correct value).

mcantelon replied:
Ah, makes sense!

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+
+    return render(request, 'administration/usage.html', locals())
+
+def _usage_dirs(calculate_usage = True):
```
No spaces around default params in a function definition please.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+
+def _usage_check_directory_used(path):
+    # Get total usage in bytes
+    du = subprocess.Popen(["du", "-b", "-c", path], stdout=subprocess.PIPE)
```
`du --bytes --summarize` returns a one-line summary of the folder, so you don't have to pipe it to tail.  Also, I prefer long options in code so it's clearer what's happening.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -33,7 +37,7 @@
 from components.administration.models import ArchivistsToolkitConfig
 from components.administration.forms import TaxonomyTermForm
-from django.http import HttpResponseRedirect
+from django.http import Http404, HttpResponseBadRequest, HttpResponseRedirect 
```
Nitpick, but don't add trailing whitespace?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+    df = subprocess.Popen(["df", path], stdout=subprocess.PIPE)
+    output = df.communicate()[0]
+    device, size, used, available, percent, mountpoint = output.split("\n")[1].split()
+    return int(size) * 512
```
[`subprocess.check_output`](https://docs.python.org/2/library/subprocess.html#subprocess.check_output) may be useful here, especially since you can catch the `CalledProcessError` and return a default if the command doesn't run properly (eg path doesn't exist).  Same advice for `_usage_check_directory_used`

Hwesta commented on src/dashboard/src/templates/administration/usage.html:
```diff
@@ -0,0 +1,49 @@
+          <td>{{ dir.size|filesizeformat }}</td>
+          <td>
+            {% if dir.description %}
+            <form method='POST' action='/administration/usage/clear/{{ id }}/'>
```
Use `{% url %}` for this instead of hard coding the path.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+            'size': shared_size,
+            'used': dip_upload_used
+        },
+        'backlog': _usage_location_info('BL', shared_size)
```
It would be nice if it was easier to add more directories to this function.  Maybe have a dict (or an OrderedDict?) with the static info for all the directories you want to display, and iterate over them and populate the size/used if calculate_usage=True?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+    # Get currently processing location
+    locations = storage_service.get_location(purpose=purpose)
+    location = locations[0]
+    if location['quota'] == None:
```
Nitpick, but `is None`

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+        messages.info(request, 'Cleared.')
+        return redirect('components.administration.views.usage')
+    else:
+        return HttpResponseBadRequest()
```
I think you want [HttpResponseNotAllowed](https://docs.djangoproject.com/en/1.7/ref/request-response/#django.http.HttpResponseNotAllowed)

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+def usage_clear(request, dir_id):
+    if request.method == 'POST':
+        usage_dirs = _usage_dirs(False)
+        dir = usage_dirs[dir_id]
```
Don't shadow builtins like `dir`

Hwesta commented on src/dashboard/src/components/administration/urls.py:
```diff
@@ -28,6 +28,8 @@
     (r'sources/$', 'sources'),
     (r'storage/$', 'storage'),
+    (r'usage/$', 'usage'),
+    (r'usage/clear/(?P<dir_id>\w+)/$', 'usage_clear'),
```
Should this be in api/ instead of administration?  Do we care about login-required vs api-key?

mcantelon replied:
Maybe it should be available both in the UI and as an API? It's in the UI so folks using AM in the cloud can do some basic housekeeping without SSH access, but it definitely would make sense to have accessible via the API for automation.

Hwesta replied:
I agree - being able to do automation would be useful, but do we want to give remote access to deleting things?  Especially for uploadedDIPs, which may contain DIPs that (despite the name :P) haven't actually been sent elsewhere and are awaiting manual download.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+            raise Http404
+
+        # Determine which directories need to be cleared
+        if 'purpose' in dir and dir['purpose'] == 'BL':
```
`if dir_info.get('purpose') == 'BL':`?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+        # Determine which directories need to be cleared
+        if 'purpose' in dir and dir['purpose'] == 'BL':
+            # Get location data
+            locations = storage_service.get_location(purpose=dir['purpose'])
```
Isn't the path already in usage_dirs?  Why re-fetch it?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+                        shutil.rmtree(entry_path)
+                except Exception, e:
+                    messages.error(request, str(e))
+                    return redirect('components.administration.views.usage')
```
Do we want to bail if _any_ file/folder fails?  Or should we just log it and continue?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +227,127 @@ def storage(request):
+                    messages.error(request, str(e))
+                    return redirect('components.administration.views.usage')
+
+        messages.info(request, 'Cleared.')
```
Would be nice if this displayed the path(s) that were deleted, for better usability.

Hwesta commented:
Do we want to be deleting the backlog?  I'm not sure that should be so easily cleared, especially since it may be the only copy of archival files halfway through being processed.

We probably also want to display info for and clear the failed, rejected, tmp and staging dirs.

mcantelon committed:
Naming exceptions, add docstrings

mcantelon commented:
I think I've implemented all recommendations (except adding a REST API interface for clearing which I can do once the admin UI way seems good to you).

mcantelon commented:
I also did some other things to get rid of hardcoding so specifying what can be cleared is fully configuration-driven. We could conceivably change things so this is read from an /etc file (YAML or something) to give sys admins the ability to customize their installation in this respect.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    except subprocess.CalledProcessError:
+        return 0
+
+def _usage_shared_dir_path():
```
Remove this and replace with a call to `helpers.get_client_config_value`?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+            # If contained, make path absolute
+            space = dirs[directory]['contained_by']
+            absolute_path = os.path.join(dirs[space]['path'], dirs[directory]['path'])
+            dirs[directory]['path'] = absolute_path
```
More stylistic, but why do this in a loop, instead of in the definition? The shared directory path should be the same for all of them, so another option is to define a shared_dir variable, and os.path.join it to everything.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    for directory in dirs:
+        if dirs[directory].get('location_purpose') is not None:
+            # If a location, determine path
+            dirs[directory]['path'] = _usage_location_path(dirs[directory]['location_purpose'])
```
What's the use case for this?  None of the above directories have 'location_purpose' defined.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+
+    # Calculate usage/size, if requested
+    if calculate_usage:
+        for directory in dirs:
```
You may find [`.iteritems()`](https://docs.python.org/2/library/stdtypes.html#dict.iteritems) useful instead of `dirs[directory]`  Modifying the value returned by the iterator modifies it in the dict.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    # Calculate usage/size, if requested
+    if calculate_usage:
+        for directory in dirs:
+            if dirs[directory].get('contained_by') is None:
```
Do you care if it's specifically None?  Or can you do `if 'contained_by' in dirs[directory]`?  Same advice applies elsewhere.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+        output = subprocess.check_output(["df", path])
+        device, size, used, available, percent, mountpoint = output.split("\n")[1].split()
+        return int(size) * 512
+    except subprocess.CalledProcessError:
```
Can this also generate other types of exceptions?  Do we want to log something if an exception is generated?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    except subprocess.CalledProcessError:
+        return 0
+
+def _usage_check_directory_used(path):
```
Python convention uses [docstrings](https://www.python.org/dev/peps/pep-0257/) to summarize what a function does at a high level, and comments for descriptions inside a function.  Since this looks like a summary, should be triple quoted.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    # Get volume size (in 512 byte blocks)
+    try:
+        output = subprocess.check_output(["df", path])
+        device, size, used, available, percent, mountpoint = output.split("\n")[1].split()
```
Python convention uses `_` to indicated [a variable we don't care about](http://stackoverflow.com/questions/5893163/what-is-the-purpose-of-the-single-underscore-variable-in-python).  If that gets ugly, you could also index into the list for the value you want (and use a comment to explain why).  Or leave it the way it is as a form of explanation.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+    # look up the path) should be specified
+    #
+    # If only certain sudirectories within a path should be deleted, set
+    # 'subdirectories' to a list of them
```
Is this still true?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+def usage(request):
+    usage_dirs = _usage_dirs()
+
+    return render(request, 'administration/usage.html', locals())
```
I'd like to start moving away from using `locals()` as it's bad security, and instead of explicitly name the variables passed to the template. Can you use `{'usage_dirs': usage_dirs}`?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+        if dir_info.get('subdirectories') is not None:
+            dirs_to_empty = []
+            for subdirectory in dir_info['subdirectories']:
+                dirs_to_empty.append(os.path.join(dir_info['path'], subdirectory))
```
Have you discovered the magic of [list comprehensions](https://docs.python.org/2/tutorial/datastructures.html#list-comprehensions)?  They're fantastic.  This could be `dirs_to_empty = [os.path.join(dir_info['path'], subdir) for subdir in dir_info['subdirectories']]`

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+                        shutil.rmtree(entry_path)
+                successes.append(directory)
+            except Exception, e:
+                errors.append(str(e))
```
Do we want to log these as well?  logger.exception or logger.foo(..., exc_info=True) may be useful.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+
+    return render(request, 'administration/usage.html', locals())
+
+def _usage_dirs(calculate_usage=True):
```
Nice touch putting the leading _ on all the internal-only functions.

mcantelon commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,160 @@ def storage(request):
+
+        # Determine if specific subdirectories need to be cleared, rather than
+        # whole directory
+        if 'subdirectories' in dir_info:
```
This functionality was originally added to accommodate Transfer Backlog clearing where we needed to clear the contents of two subdirectories, rather than the whole directory's contents. I'm wondering if I should leave this in in case it's useful or if I should remove it?

mcantelon replied:
(Probably removing it makes sense as it's not hard to re-add.)

mcantelon commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -223,6 +229,146 @@ def storage(request):
+
+    # Resolve location paths and make relative paths absolute
+    for _, dir_spec in dirs.iteritems():
+        if 'contained_by' in dir_spec:
```
I do the contained_by thing, rather than always path.join'ing the shared directory, in case in the future we want to support having things on multiple volumes, etc. Maybe it's overkill for now?

Hwesta replied:
Seems reasonable

mcantelon commented:
Merged.

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -184,6 +190,146 @@ def storage(request):
+        size = usage_summary.split()[1]
+
+        return int(size) * 512
+    except Exception, e:
```
The modern syntax for providing a name for an exception is [`except Exception as e`](http://stackoverflow.com/questions/5119751/in-python-whats-the-difference-between-except-exception-as-e-and-except-exc).  Also, catching `Exception` is generally a [bad practice](http://ischenko.blogspot.ca/2005/01/exception-based-code-antipatterns.html), as it silences all errors, not just the ones you're expecting.

I was thinking something like

``` python
except (ValueError, IndexError, subprocess.CalledProcessError):  # Known errors that could be raised
    logger.exception('Error checking directory size for %s', path)  # logger.exception already logs the exception raised
    return 0
```

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -184,6 +190,146 @@ def storage(request):
+    try:
+        output = subprocess.check_output(["du", "--bytes", "--summarize", path])
+        return output.split("\t")[0]
+    except Exception, e:
```
Same comment here as above.

Hwesta commented:
Can you update all your `except Exception, e: logger.exception(str(e))` to name the Exception(s) that are being caught, and provide a useful logging message? logger.exception already provides the exception it caught along with a traceback, so a human-useful error message is more appropriate.  Also, logger.exception always generates an ERROR level message - a WARNING level might be better in some cases (`logger.warning('message', exc_info=True)`) - your call.

Also, can you add docstrings to your functions? For most of them a one-liner summary is fine, but for the longer ones an explanation of what the function is trying to accomplish, restrictions on the params and return value is very useful when reading later.  We're using Sphinx's reST style, here's a sample:

``` python
"""
Short summary

Optional longer summary

:param optional-type-of-param name-of-param: Description of what param does
:param str path: Path to get the size of
:param flags: Bool for something in this function
:returns: Size of the path in bytes
"""
```

mcantelon commented:
Okay, I've added docstrings and made the exception handling less generic.

Hwesta committed:
Adminstration views: Re-add processing_views import, clean up

Hwesta committed:
Administration views: Fix docstrings, logging

Move docstrings inside the function they describe. Logger calls use built-in
logging formatting.
